### DISCUSSION TITLE
How to use external libraries (not bundled along the app)

### DISCUSSION THREAD

**Discussion by Thoufak** [pyinstaller/pyinstaller#8190] [Help] [answered]
Some of my app users don't need all the installed libraries. So, in order to reduce the exe file size, I'd like to be able to exclude those libraries from the bundle, but still be able to provide them as separate files. Therefore, my question is whether there is a way to import packages that are not bundled with the app?

**Comment by rokm** [comment, marked as answer]
If you want to load packages from external location, you need to adjust `sys.path` at the run-time; same as you would with regular python. Be aware though, that if a package is (partially) collected into the frozen application, you will not be able to load same (full) package from the external location - so you need to ensure it is completely excluded.

**Reply by Thoufak** [reply]
> Be aware though, that if a package is (partially) collected into the frozen application, you will not be able to load same (full) package from the external location - so you need to ensure it is completely excluded.

Did I understand you right that "import X" will first try to import the X from the bundled packages and only if it can't find it, it will search in the paths outside the bundle (given such paths are specified in the `sys.path`)?

**Reply by rokm** [reply]
Yes, the import will first be attempted using PyInstaller's `PyiFrozenImporter` (because we place it first in `sys.meta_paths`); if it cannot find the package, then other finders will be used (such as python's built-in `PathFinder`, which looks at locations specified in `sys.path`).

**Reply by Thoufak** [reply]
Thanks a lot!

**Reply by rokm** [reply]
And for the sake of completeness - `PyiFrozenImporter` sets the packages' locations to `sys._MEIPASS/{package_name}`. So parts of (partially) collected packages that not handled by `PyiFrozenImporter` (are not collected into PYZ archive; e.g., source-only files, binary extensions) can also be discovered by the built-in `PathFinder`, at the same location ( `sys._MEIPASS/{package_name}`). 

This is why if `package_name` was handled by `PyiFrozenImporter` (loaded from frozen application, even if partially), the contents of `package_name` in `/external/package_name` cannot be discovered even if `/external` is in `sys.path`. Because the `package_name.__path__` is set to  `[sys._MEIPASS/{package_name}]`. You could also import `package_name` and append the external location to its `__path__`, but it should be easier to just completely exclude the package (and it is also more consistent with your scenario, in which you do not need the package in frozen application at all).
